spot.go

```diff
@@ -2,65 +2,87 @@ package main
 
 import (
 	"context"
-	"io/ioutil"
-	"net/http"
+	"errors"
 	"time"
 
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+
 	log "github.com/sirupsen/logrus"
 )
 
 const (
 	metadataURLTerminationTime = "http://169.254.169.254/latest/meta-data/spot/termination-time"
-	terminationTransition      = "ec2:SPOT_INSTANCE_TERMINATION"
 	terminationTimeFormat      = "2006-01-02T15:04:05Z"
 )
 
-func pollSpotTermination(ctx context.Context) chan time.Time {
-	ch := make(chan time.Time)
-
-	log.Debugf("Polling metadata service for spot termination notices")
-
-	go func() {
-		// Close channel before returning since this (goroutine) is the sending side.
-		defer close(ch)
-		retry := time.NewTicker(time.Second * 5).C
-	Loop:
-		for {
-			select {
-			case <-ctx.Done():
-				break Loop
-			case <-retry:
-				res, err := http.Get(metadataURLTerminationTime)
-				if err != nil {
-					log.WithError(err).Info("Failed to query metadata service")
-					continue
-				}
-
-				// We read the body immediately so that we can close the body in one place
-				// and still use 'continue' if any of our conditions are false.
-				body, err := ioutil.ReadAll(res.Body)
-				res.Body.Close()
-				if err != nil {
-					log.WithError(err).Info("Failed to read response from metadata service")
-					continue
-				}
+// NewSpotListener ...
+func NewSpotListener(instanceID string, metadata *ec2metadata.EC2Metadata) *SpotListener {
+	return &SpotListener{
+		listenerType: "spot",
+		instanceID:   instanceID,
+		metadata:     metadata,
+	}
+}
 
-				// will return 200 OK with termination notice
-				if res.StatusCode != http.StatusOK {
-					continue
-				}
+// SpotListener ...
+type SpotListener struct {
+	listenerType string
+	instanceID   string
+	metadata     *ec2metadata.EC2Metadata
+}
 
-				// if 200 OK, expect a body like 2015-01-05T18:02:00Z
-				t, err := time.Parse(terminationTimeFormat, string(body))
-				if err != nil {
-					log.WithError(err).Info("Failed to parse time in termination notice")
-					continue
-				}
+// Type returns a string describing the listener type.
+func (l *SpotListener) Type() string {
+	return l.listenerType
+}
 
-				ch <- t
+// Start the spot termination notice listener.
+func (l *SpotListener) Start(ctx context.Context, notices chan<- Notice) error {
+	if !l.metadata.Available() {
+		return errors.New("ec2 metadata is not available")
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.NewTicker(time.Second * 5).C:
+			log.Debugf("Polling ec2 metadata for spot termination notices")
+			out, err := l.metadata.GetMetadata("spot/termination-time")
+			if err != nil {
+				log.WithError(err).Error("Failed to get spot termination")
```

Comment:
Note that this will fire every time there isn't a spot termination. This was what caused me problems with my implementation of it. I couldn't figure out how to unwrap there error to figure out if it was a 404 error (meaning no termination time) or some other type of error.

Reply:
Good catch, I'll look into it and see if there is a way to check the error properly!